test_util/cluster_api.py

```diff
@@ -241,280 +241,43 @@ def __init__(self, dcos_uri, masters, public_masters, slaves, public_slaves,
         # Make URI never end with /
         self.dcos_uri = dcos_uri.rstrip('/')
 
-    def cluster_request(self, request_method, path, node=None, port=None, **kwargs):
-        """Args:
-            request_method: (str) name of method for requests.request
-            path: (str) URL path to request
-            node: (str) the hostname of the node to be requested from, if node=None,
-                then public cluster address (see environment DCOS_DNS_ADDRESS)
-            port: (int) port to be requested at. If port=None, the adminrouter port
-                for that given node type will be used
-            user: (test_util.helpers.DcosUser) if this argument is specified, this
-                user's auth header will be used to make the request. If user=None,
-                then the web_auth_default_user's auth will be used. If user=None,
-                then no auth headers are used, even if auth is enabled
-            **kwargs: any optional arguments to be passed to requests.request
-        """
-        assert request_method in self.request_methods
-        user = kwargs.pop('user', self.web_auth_default_user)
-        if node is None:
-            node = self.dns_host
-        headers = kwargs.pop('headers', {})
-        if self.auth_enabled and user:
-            headers.update(user.auth_header)
-        url = self.get_url(node=node, path=path, port=port)
-        logging.info('{}: {}'.format(request_method, url))
-        return requests.request(request_method, url, headers=headers, **kwargs)
-
-    def get_url(self, node, path, port=None):
-        """Looks up node type and appends the appropriate port and scheme
-        """
-        if node in self.masters or node == self.dns_host:
-            default_port = self.adminrouter_master_port[self.scheme]
-        elif node in self.all_slaves:
-            default_port = self.adminrouter_agent_port[self.scheme]
+    def get_user_session(self, user):
+        new_session = copy.deepcopy(self)
```

Comment:
The only thing that changes is the web_auth_default_user. The magic state lies in whether the user has been authenticated yet, which is resolved here by doing that first. I could instead do:
```
if user:
    user.authenticate(self)
return self.__class__(<pass all the necessary args w/ user as the new web_auth_default_user>)
```
but the reason I didnt opt for that is I liked how clean it looked to do copy.deepcopy instead of having to set 11 args